pyadjoint/tape.py

```diff
@@ -193,6 +194,18 @@ def copy(self):
         # TODO: Offer deepcopying. But is it feasible memory wise to copy all checkpoints?
         return Tape(blocks=self._blocks[:])
 
+    def block_vars(self, tag=None):
+        """Return a dictionary mapping all block variables on the tape to their checkpointed values"""
+
+        return {
+            var: var.checkpoint
+            for var in chain.from_iterable(
+                chain(
+                    b.get_dependencies(),
+                    b.get_outputs()
```

Comment:
I think this should only capture outputs. The reason is that anything which is only a dependency and not an output is actually an input to the tape. Its checkpoint should be left in place to be used on the next tape evaluation (and can safely be left on the tape because it won't be overwritten).

Reply:
I think the problem with that as written is that there is at least one dependency to update: the control variable. I don't know of any way to update that explicitly, since it would normally only be set when evaluating the ReducedFunctional.

Reply:
True, but that doesn't mean we get to throw out all the other inputs. I think your argument means that we need all outputs and inputs which are also controls (the tape has a mechanism for marking controls).

Reply:
Makes sense! I've changed the method to accept a list of controls, from which we extract the `.block_variable` (like elsewhere in the tape methods).